File: tests/libs/hdf5/test-env-variables/1_8/FORTRAN/H5G/h5ex_g_compact.f90
!************************************************************
!
!  This example shows how to create "compact-or-indexed"
!  format groups, new to 1.8.  This example also illustrates
!  the space savings of compact groups by creating 2 files
!  which are identical except for the group format, and
!  displaying the file size of each.  Both files have one
!  empty group in the root group.
!
!  This file is intended for use with HDF5 Library verion 1.8
!
!************************************************************

PROGRAM main

  USE HDF5
  IMPLICIT NONE

  CHARACTER(LEN=18), PARAMETER :: filename1 = "h5ex_g_compact1.h5"
  CHARACTER(LEN=18), PARAMETER :: filename2 = "h5ex_g_compact2.h5"
  CHARACTER(LEN=2) , PARAMETER :: groupname  = "G1"

  INTEGER(HID_T)   :: file, group, fapl ! handles
  INTEGER :: hdferr
  INTEGER :: storage_type ! Type of storage for links in group:
                          !   H5G_STORAGE_TYPE_COMPACT: Compact storage
                          !   H5G_STORAGE_TYPE_DENSE: Indexed storage
                          !   H5G_STORAGE_TYPE_SYMBOL_TABLE: Symbol tables
  INTEGER :: nlinks       ! Number of links in group
  INTEGER :: max_corder   ! Current maximum creation order value for group 
  INTEGER(HSIZE_T) :: size
  !
  ! Initialize FORTRAN interface.
  !
  CALL h5open_f(hdferr)
  !
  ! Create file 1.  This file will use original format groups.
  !
  CALL h5fcreate_f(filename1, H5F_ACC_TRUNC_F, file, hdferr)
  CALL h5gcreate_f(file, groupname, group, hdferr)
  !
  ! Obtain the group info and print the group storage type.
  !
  CALL H5Gget_info_f(group, storage_type, nlinks, max_corder, hdferr)

  WRITE(*,'("Group storage type for ",A," is: ")', ADVANCE='NO' ) filename1

  IF(storage_type.EQ.H5G_STORAGE_TYPE_COMPACT_F)THEN
     WRITE(*,'("H5G_STORAGE_TYPE_COMPACT_F")') ! New compact format
  ELSE IF(storage_type.EQ.H5G_STORAGE_TYPE_DENSE_F)THEN
     WRITE(*,'("H5G_STORAGE_TYPE_DENSE_F")') ! New dense (indexed) format
  ELSE IF(storage_type.EQ.H5G_STORAGE_TYPE_SYMBOL_TABLE_F)THEN
     WRITE(*,'("H5G_STORAGE_TYPE_SYMBOL_TABLE_F")') ! Original format
  ENDIF
  !
  ! Close and re-open file.  Needed to get the correct file size.
  !
  CALL h5gclose_f(group, hdferr)
  CALL h5fclose_f(file , hdferr)
  CALL h5fopen_f(filename1, H5F_ACC_RDONLY_F, file, hdferr)
  !
  ! Obtain and print the file size.
  !
  CALL h5fget_filesize_f(file, size, hdferr)
  WRITE(*,'("File size for ",A," is: ",i6," bytes",/)') filename1, size
  !
  ! Close filename1
  !
  CALL h5fclose_f(file, hdferr)
  !
  ! Set file access property list to allow the latest file format.
  ! This will allow the library to create new compact format groups.
  !
  CALL h5pcreate_f (H5P_FILE_ACCESS_F, fapl, hdferr)
  CALL h5pset_libver_bounds_f (fapl, H5F_LIBVER_LATEST_F, H5F_LIBVER_LATEST_F, hdferr)
  !
  ! Create file 2 using the new file access property list.
  !
  CALL h5fcreate_f(filename2, H5F_ACC_TRUNC_F, file, hdferr, access_prp=fapl)
  CALL h5gcreate_f(file, groupname, group, hdferr)
  !
  ! Obtain the group info and print the group storage type.
  !
  CALL H5Gget_info_f(group, storage_type, nlinks, max_corder, hdferr)

  WRITE(*,'("Group storage type for ",A," is: ")', ADVANCE='NO' ) filename2

  IF(storage_type.EQ.H5G_STORAGE_TYPE_COMPACT_F)THEN
     WRITE(*,'("H5G_STORAGE_TYPE_COMPACT_F")') ! New compact format
  ELSE IF(storage_type.EQ.H5G_STORAGE_TYPE_DENSE_F)THEN
     WRITE(*,'("H5G_STORAGE_TYPE_DENSE_F")') ! New dense (indexed) format
  ELSE IF(storage_type.EQ.H5G_STORAGE_TYPE_SYMBOL_TABLE_F)THEN
     WRITE(*,'("H5G_STORAGE_TYPE_SYMBOL_TABLE_F")') ! Original format
  ENDIF
  !
  ! Close and re-open file.  Needed to get the correct file size.
  !
  CALL h5gclose_f(group, hdferr)
  CALL h5fclose_f(file , hdferr)
  CALL h5fopen_f(filename2, H5F_ACC_RDONLY_F, file, hdferr, fapl)
  !
  ! Obtain and print the file size.
  !
  CALL h5fget_filesize_f(file, size, hdferr)
  WRITE(*,'("File size for ",A," is: ",i6,"bytes",/)') filename2, size
  !
  ! Close and release resources.
  !
  CALL h5pclose_f(fapl, hdferr)
  CALL h5fclose_f(file, hdferr)

END PROGRAM main

File: tests/libs/hdf5/test-env-variables-new/C/H5G/h5ex_g_compact.c
/************************************************************

  This example shows how to create "compact-or-indexed"
  format groups, new to 1.8.  This example also illustrates
  the space savings of compact groups by creating 2 files
  which are identical except for the group format, and
  displaying the file size of each.  Both files have one
  empty group in the root group.

  This file is intended for use with HDF5 Library version 1.8

 ************************************************************/

#include "hdf5.h"
#include <stdio.h>

#define FILE1       "h5ex_g_compact1.h5"
#define FILE2       "h5ex_g_compact2.h5"
#define GROUP       "G1"

int
main (void)
{
    hid_t       file, group, fapl;         /* Handles */
    herr_t      status;
    H5G_info_t  ginfo;
    hsize_t     size;

    /*
     * Create file 1.  This file will use original format groups.
     */
    file = H5Fcreate (FILE1, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
    group = H5Gcreate (file, GROUP, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);

    /*
     * Obtain the group info and print the group storage type.
     */
    status = H5Gget_info (group, &ginfo);
    printf ("Group storage type for %s is: ", FILE1);
    switch (ginfo.storage_type) {
        case H5G_STORAGE_TYPE_COMPACT:
            printf ("H5G_STORAGE_TYPE_COMPACT\n"); /* New compact format */
            break;
        case H5G_STORAGE_TYPE_DENSE:
            printf ("H5G_STORAGE_TYPE_DENSE\n"); /* New dense (indexed) format */
            break;
        case H5G_STORAGE_TYPE_SYMBOL_TABLE:
            printf ("H5G_STORAGE_TYPE_SYMBOL_TABLE\n"); /* Original format */
            break;
        case H5G_STORAGE_TYPE_UNKNOWN:
            printf ("H5G_STORAGE_TYPE_UNKNOWN\n"); /* Unknown format */
    }

    /*
     * Close and re-open file.  Needed to get the correct file size.
     */
    status = H5Gclose (group);
    status = H5Fclose (file);
    file = H5Fopen (FILE1, H5F_ACC_RDONLY, H5P_DEFAULT);

    /*
     * Obtain and print the file size.
     */
    status = H5Fget_filesize (file, &size);
    printf ("File size for %s is: %d bytes\n\n", FILE1, (int)size);

    /*
     * Close FILE1.
     */
    status = H5Fclose (file);

    /*
     * Set file access property list to allow the latest file format.
     * This will allow the library to create new compact format groups.
     */
    fapl = H5Pcreate (H5P_FILE_ACCESS);
    status = H5Pset_libver_bounds (fapl, H5F_LIBVER_LATEST, H5F_LIBVER_LATEST);

    /*
     * Create file 2 using the new file access property list.
     */
    file = H5Fcreate (FILE2, H5F_ACC_TRUNC, H5P_DEFAULT, fapl );
    group = H5Gcreate (file, GROUP, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);

    /*
     * Obtain the group info and print the group storage type.
     */
    status = H5Gget_info (group, &ginfo);
    printf ("Group storage type for %s is: ", FILE2);
    switch (ginfo.storage_type) {
        case H5G_STORAGE_TYPE_COMPACT:
            printf ("H5G_STORAGE_TYPE_COMPACT\n"); /* New compact format */
            break;
        case H5G_STORAGE_TYPE_DENSE:
            printf ("H5G_STORAGE_TYPE_DENSE\n"); /* New dense (indexed) format */
            break;
        case H5G_STORAGE_TYPE_SYMBOL_TABLE:
            printf ("H5G_STORAGE_TYPE_SYMBOL_TABLE\n"); /* Original format */
            break;
        case H5G_STORAGE_TYPE_UNKNOWN:
            printf ("H5G_STORAGE_TYPE_UNKNOWN\n"); /* Unknown format */
    }

    /*
     * Close and re-open file.  Needed to get the correct file size.
     */
    status = H5Gclose (group);
    status = H5Fclose (file);
    file = H5Fopen (FILE2, H5F_ACC_RDONLY, fapl);

    /*
     * Obtain and print the file size.
     */
    status = H5Fget_filesize (file, &size);
    printf ("File size for %s is: %d bytes\n", FILE2, (int)size);
    printf ("\n");

    /*
     * Close and release resources.
     */
    status = H5Pclose (fapl);
    status = H5Fclose (file);

    return 0;
}
